### PR TITLE
Feature/ide url handler

### PR DESCRIPTION
**Pull request by Jibbarth**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | #237

Thanks to symfony/console 4.3, we can add hyperlink easily in console. 

This PR add a `ide` config key, (fillable by theses differents editor: 
phpstorm, sublime, textmate, macvim, emacs, atom and vscode) and generate hyperlinks in terminal to file that have issues.

### CODE REVIEW

File: docs/ide.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+In your `phpinsights.php` file, add the config `'ide' => 'myide'`.
+
+For example:
+
+```php
+<?php
+
+return [
+    'ide' => 'vscode',
```
Comment (olivernybroe) [inline docs/ide.md]: Can you add a `...` before and after?

File: docs/ide.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+that contains an `%f` placeholder where the file path is expected and `%l` 
+placeholder for the line number.
+
+For example:
+
+```php
+<?php
+
+return [
+    'ide' => 'myide://open?url=file://%f&line=%l',
```
Comment (olivernybroe) [inline docs/ide.md]: This is a really nice feature!

File: src/Application/Console/Analyser.php [hunk 3]
```diff
@@ -52,6 +52,10 @@ public function analyse(
         $insightCollection = $this->insightCollectionFactory
             ->get($metrics, $config, $dir, $consoleOutput);
 
+        if (method_exists($formatter, 'injectFileLinkFormatter')) {
```
Comment (olivernybroe) [inline src/Application/Console/Analyser.php]: Not sure if I like this approach. I feel like it doesn't belong here. Could we just give the format method the config file or the config file in the constructor of the formatter?

Comment (Jibbarth) [reply src/Application/Console/Analyser.php]: Yeah, I'm not convinced either. 

I initially added it in format function, but it's not a prerequisite for each formatter, that's why I think adding it in the constructor it's not a good solution... 

Maybe by create an Interface `RequireFileLinkFormatter` ? And adding it if formatter is instance of this interface. 
WDYT @olivernybroe ?

Comment (olivernybroe) [reply src/Application/Console/Analyser.php]: Hmm, still seems kinda weird as that is only needed for that formatter.

What if we in the constructor of the Console formatter just extract the parameter `ide` from the `inputInteface`? This way all the logic is in the `Console` formatter, but no extra parameter is needed.

File: docs/ide.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# IDE Integration
```
Comment (nunomaduro) [inline docs/ide.md]: I would remove this, and add only:
```
# IDE Integration

In your `phpinsights.php` file, add the conf...

## Supported IDE 
...
## Unsupported IDE
...
```

File: src/Application/ConfigResolver.php [hunk 5]
```diff
@@ -109,4 +115,48 @@ public static function mergeConfig(array $base, array $replacement): array
...
+            throw new \LogicException(sprintf(
+                'Unknow IDE "%s". Try one in this list [%s] or provide pattern link handler',
+                $ide,
+                implode(', ', array_keys($links))
+            ));
+        }
+
+        $fileFormatterPattern = $links[$ide] ?? $ide;
+
+        $fileLinkFormatter = new NullFileLinkFormatter();
```
Comment (nunomaduro) [inline src/Application/ConfigResolver.php]: Can you make a ternary here and return directly the result?

```php

return $fileFormatterPattern === null ? new NullFileLinkFormatter() : new FileLinkFormatter($fileFormatterPattern);
```

File: stubs/config.php [hunk 6]
```diff
@@ -19,6 +19,23 @@
 
     'preset' => 'default',
 
+
+    /*
+    |--------------------------------------------------------------------------
+    | IDE
+    |--------------------------------------------------------------------------
+    |
+    | This options allow to add hyperlinks in your terminal to quickly open
```
Comment (nunomaduro) [inline stubs/config.php]: Can you make this comments just like in Laravel? Every paragraph is 3 chars shorter then the last one.

File: docs/ide.md (hunk 4 above)
Comment (Jibbarth) [reply docs/ide.md]: Do you mean without the **prerequisite & troubleshouting** part ? 

Should i display them as [tip or warning](https://v1.vuepress.vuejs.org/guide/markdown.html#custom-containers) instead ?

File: stubs/config.php (hunk 6 above)
Comment (Jibbarth) [reply stubs/config.php]: :roll_eyes:  Waw ! I didn't pay attention to that ! It's quite hard :sweat_smile: 

Is it better now ?

File: src/Application/Console/Analyser.php (hunk 3 above)
Comment (Jibbarth) [reply src/Application/Console/Analyser.php]: In fact, it's used by the Console formatter, but we can imagine use it also in an HTML Formatter. That's why I would prefer keeping the ide resolving outside the Console Formatter.

Maybe the easier would be to create an Config class, that will be instancied when the app boots, and dispatch it through the container ? Then each class that need a part of config can get it from container ?

Comment (olivernybroe) [reply src/Application/Console/Analyser.php]: I like that idea. That would work great for when we add more features. 

Another idea could be to use a trait, but I'll let you choose, I like both ideas.

File: src/Application/Console/Formatters/Console.php [hunk 7]
```diff
@@ -32,9 +36,28 @@ final class Console implements Formatter
...
+
+    /**
+     * @var bool
+     */
+    private $supportHyperLinks;
+
     public function __construct(InputInterface $input, OutputInterface $output)
     {
         $this->style = new Style($input, $output);
+         false;
```
Comment (nunomaduro) [inline src/Application/Console/Formatters/Console.php]: What this false is about?

File: src/Domain/Details.php [hunk 8]
```diff
@@ -73,6 +73,9 @@ public function hasFile(): bool
 
     public function getLine(): int
     {
+        if (! $this->hasLine()) {
```
Comment (nunomaduro) [inline src/Domain/Details.php]: Can we use a ternary here?

**Comment by olivernybroe** [comment]
@Jibbarth When the config file PR is merged in, are you then going to update this PR also?

**Comment by Jibbarth** [comment]
@olivernybroe Sure :blush:

**Pull request merged**